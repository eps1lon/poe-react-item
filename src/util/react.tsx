import * as React from 'react';

// text nodes dont need keys
export const CommaSeparator = () => ', ';

export const LineBreak = (key: string) => <br key={key} />;

export function intersperse(
  nodes: React.ReactNode[],
  renderSeparator: (key: string) => React.ReactNode = CommaSeparator,
) {
  if (nodes.length === 0) {
    return [];
  }

  return nodes.slice(1).reduce(
    (joined: React.ReactNode[], element, index) => {
      // push instead of concat's returnval because concat is expensive
      joined.push(renderSeparator(`separator-${index}`), element);
      return joined;
    },
    [nodes[0]],
  );
}
