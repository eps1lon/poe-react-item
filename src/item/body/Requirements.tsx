import * as React from 'react';

import { Requirements } from '../poe';
import PropertyValue, { DisplayPropertyType } from './PropertyValue';
import { intersperse, CommaSeparator } from '../../util/react';
import { AugmentableValue, SingleValue } from '../../util/value';

export interface Props {
  requirements: Requirements;
}

export default class RequirementsComponent extends React.PureComponent<Props> {
  public render() {
    if (!this.hasAny()) {
      return null;
    }

    return (
      <div className="requirements">
        Requires {intersperse(this.requirements(), CommaSeparator)}
      </div>
    );
  }

  private hasAny() {
    const { requirements } = this.props;
    // Object.values throws ts error (prop not found)
    return Object.keys(requirements).filter(key => {
      const requirement = requirements[key];
      return requirement !== undefined && requirement.value > 0;
    });
  }

  private requirements() {
    const {
      requirements: { level, strength, dexterity, inteliligence },
    } = this.props;

    return [
      this.requirement('level', level),
      this.attributeRequirement('str', strength),
      this.attributeRequirement('dex', dexterity),
      this.attributeRequirement('int', inteliligence),
    ].filter(Boolean);
  }

  private requirement(
    human: string,
    augmentable?: AugmentableValue<SingleValue>,
  ) {
    if (augmentable === undefined || augmentable.value < 1) {
      return null;
    }

    return [
      <span key="human">{human}</span>,
      ' ',
      <PropertyValue
        value={String(augmentable.value)}
        type={
          augmentable.augmented
            ? DisplayPropertyType.augmented
            : DisplayPropertyType.default
        }
      />,
    ];
  }

  private attributeRequirement(
    human: string,
    augmentable?: AugmentableValue<SingleValue>,
  ) {
    if (augmentable === undefined || augmentable.value < 1) {
      return null;
    }

    return [
      <PropertyValue
        value={String(augmentable.value)}
        type={
          augmentable.augmented
            ? DisplayPropertyType.augmented
            : DisplayPropertyType.default
        }
      />,
      ' ',
      <span key="human">{human}</span>,
    ];
  }
}
