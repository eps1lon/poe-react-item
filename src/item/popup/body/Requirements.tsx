import * as React from 'react';

import { Requirements } from '../../poe';
import PropertyValue, { DisplayPropertyType } from './PropertyValue';
import { intersperse, CommaSeparator } from '../../../util/react';
import { AugmentableValue, SingleValue } from '../../../util/value';

export interface Props {
  requirements: Requirements;
}

export default class RequirementsComponent extends React.PureComponent<Props> {
  public static hasAny(requirements: Props['requirements']) {
    // Object.values throws ts error (prop not found)
    return (
      Object.keys(requirements).filter(key => {
        const requirement = requirements[key];
        return requirement !== undefined && requirement.value > 0;
      }).length > 0
    );
  }

  public render() {
    if (!this.hasAny()) {
      return null;
    }

    return (
      <div className="requirements">
        Requires {intersperse(this.requirements(), CommaSeparator)}
      </div>
    );
  }

  private hasAny() {
    return RequirementsComponent.hasAny(this.props.requirements);
  }

  private requirements() {
    const {
      requirements: { level, strength, dexterity, intelligence },
    } = this.props;

    return [
      this.requirement('level', level),
      this.attributeRequirement('str', strength),
      this.attributeRequirement('dex', dexterity),
      this.attributeRequirement('int', intelligence),
    ].filter(Boolean);
  }

  private requirement(
    human: string,
    augmentable?: AugmentableValue<SingleValue>,
  ) {
    if (augmentable === undefined || augmentable.value < 1) {
      return null;
    }

    return [
      <span key="human">{human}</span>,
      ' ',
      <PropertyValue
        key="value"
        value={String(augmentable.value)}
        type={
          augmentable.augmented
            ? DisplayPropertyType.augmented
            : DisplayPropertyType.simple
        }
      />,
    ];
  }

  private attributeRequirement(
    human: string,
    augmentable?: AugmentableValue<SingleValue>,
  ) {
    if (augmentable === undefined || augmentable.value < 1) {
      return null;
    }

    return [
      <PropertyValue
        key="value"
        value={String(augmentable.value)}
        type={
          augmentable.augmented
            ? DisplayPropertyType.augmented
            : DisplayPropertyType.simple
        }
      />,
      ' ',
      <span key="human">{human}</span>,
    ];
  }
}
