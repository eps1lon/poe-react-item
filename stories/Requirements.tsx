import * as classnames from 'classnames';
import * as React from 'react';
import { storiesOf } from '@storybook/react';

import Requirements, {
  Props as RequirementsProps,
} from '../src/item/popup/body/Requirements';

export interface Props extends RequirementsProps {
  rarity: 'normal' | 'magic' | 'rare';
}

const Story: React.SFC<Props> = ({ rarity, ...props }) => {
  return (
    <div className={classnames('poe-item', rarity)}>
      <section className="body">
        <Requirements {...props} />
      </section>
    </div>
  );
};

storiesOf('requirements', module)
  .add('all', () => (
    <Story
      rarity="rare"
      requirements={{
        level: { value: 5 },
        dexterity: { value: 10 },
        inteliligence: { value: 1 },
        strength: { value: 500, augmented: true },
      }}
    />
  ))
  .add('once', () => (
    <Story
      rarity="rare"
      requirements={{
        level: { value: 5 },
      }}
    />
  ));
