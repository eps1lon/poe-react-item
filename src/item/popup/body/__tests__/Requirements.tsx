import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Requirements from '../Requirements';

it('should display all', () => {
  expect(
    renderer
      .create(
        <Requirements
          requirements={{
            level: { value: 5 },
            dexterity: { value: 10 },
            intelligence: { value: 1 },
            strength: { value: 500, augmented: true },
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});

it('should display once', () => {
  expect(
    renderer
      .create(
        <Requirements
          requirements={{
            level: { value: 5 },
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});
