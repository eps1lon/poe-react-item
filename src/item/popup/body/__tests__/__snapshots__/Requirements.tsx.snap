// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display all 1`] = `
<div
  className="requirements"
>
  Requires 
  <span>
    level
  </span>
   
  <span
    className="value default"
  >
    5
  </span>
  , 
  <span
    className="value augmented"
  >
    500
  </span>
   
  <span>
    str
  </span>
  , 
  <span
    className="value default"
  >
    10
  </span>
   
  <span>
    dex
  </span>
  , 
  <span
    className="value default"
  >
    1
  </span>
   
  <span>
    int
  </span>
</div>
`;

exports[`should display once 1`] = `
<div
  className="requirements"
>
  Requires 
  <span>
    level
  </span>
   
  <span
    className="value default"
  >
    5
  </span>
</div>
`;
